### PR TITLE
Optimise hookshot and audit log backups and restores and MySQL restores

### PR DESCRIPTION
**Pull request by lildude**
Whilst examining the code for something unrelated, I noticed a few places where the uncompressed data is sent across the network rather than the compressed data, and these locations are all for data that compresses well: hookshot and audit logs and MySQL.

This PR addresses that by...

- compressing the audit and hookshot logs on the appliance side before sending across the network during the backup
- transferring the compressed audit and hookshot logs and MySQL backups across the network and then uncompressing on the appliance at the time of import.

Overall, very small instances may see a slight increase in restore times but this shouldn't be significant and will "disappear" when the amount of data increases.  Larger instances should see an improvement. Both scenarios should see more consistent timings for the hookshot logs, audit logs and MySQL steps.

---

## Testing benchmarks

All tests were performed five times. In the case of backups, the `data` directory was removed between each to ensure the full transfer occurred each time.

### Audit Log and Hookshot Backup Before:
```
ghe-backup-es-audit-log took 33s
ghe-backup-es-audit-log took 117s
ghe-backup-es-audit-log took 28s
ghe-backup-es-audit-log took 56s
ghe-backup-es-audit-log took 28s
```
```
ghe-backup-es-hookshot took 188s
ghe-backup-es-hookshot took 514s
ghe-backup-es-hookshot took 145s
ghe-backup-es-hookshot took 80s
ghe-backup-es-hookshot took 105s
```

Average times: 52.4s and 206.4s respectively (36.25s / 129.5s excluding anomaly).

### Audit Log and Hookshot Backup After:
```
ghe-backup-es-audit-log took 56s
ghe-backup-es-audit-log took 27s
ghe-backup-es-audit-log took 62s
ghe-backup-es-audit-log took 28s
ghe-backup-es-audit-log took 27s
```
```
ghe-backup-es-hookshot took 69s
ghe-backup-es-hookshot took 57s
ghe-backup-es-hookshot took 102s
ghe-backup-es-hookshot took 56s
ghe-backup-es-hookshot took 58s
```

Average times: 40s and 68.4s respectively.

### MySQL Restore Before:
```
ghe-import-mysql took 78s
ghe-import-mysql took 120s
ghe-import-mysql took 57s
ghe-import-mysql took 59s
ghe-import-mysql took 66s
```

Average time: 76s

### MySQL Restore After:
```
ghe-restore-mysql took 52s
ghe-restore-mysql took 51s
ghe-restore-mysql took 64s
ghe-restore-mysql took 52s
ghe-restore-mysql took 52s
```

Average time: 54.2s

For reference, using rsync to transfer the MySQL file took a little more time than `cat | dd`, which is expected due to the additional overhead rsync introduces:
```
ghe-restore-mysql took 60s
ghe-restore-mysql took 60s
ghe-restore-mysql took 59s
ghe-restore-mysql took 59s
ghe-restore-mysql took 59s
```

Average time: 59.4s

Overall, smaller datasets may see a slight increase in time, but things should be more consistent. Larger datasets should see a noticeable improvement.

### CODE REVIEW

**Comment by lildude** [comment]
I experimented with making `ghe-import-mysql` accept gzip compressed data input and this shaves off a bit more time:

```
ghe-restore-mysql took 46s
ghe-restore-mysql took 46s
ghe-restore-mysql took 45s
ghe-restore-mysql took 46s
ghe-restore-mysql took 46s
```

This additional improvement would then be specific to certain versions of GitHub Enterprise that contain the change.

File: share/github-backup-utils/ghe-restore-es-audit-log [hunk 1]
```diff
@@ -43,22 +43,31 @@ fi
...
+
+  if $CLUSTER || [ -n "$configured" ]; then
+    for index in $(cat $tmp_list | sed 's/\.gz$//g'); do
+      ghe_verbose "* Restoring $index"
+      echo "export PATH=\$PATH:/usr/local/share/enterprise && gzip -dc $GHE_REMOTE_DATA_USER_DIR/elasticsearch-restore/$index | ghe-es-load-json 'http://localhost:9201/$index'" |
+      ghe-ssh "$GHE_HOSTNAME" -- /bin/bash 1>&3
+    done
+  fi
+
+  ghe-ssh "$GHE_HOSTNAME" -- "sudo rm $GHE_REMOTE_DATA_USER_DIR/elasticsearch-restore/*.gz" 1>&3
```
Comment (snh) [inline share/github-backup-utils/ghe-restore-es-audit-log]: Doesn't this `rm` need to happen as part of the `if $CLUSTER || [ -n "$configured" ];` block?

File: share/github-backup-utils/ghe-restore-es-hookshot [hunk 2]
```diff
@@ -30,22 +30,30 @@ fi
...
+
+  if $CLUSTER || [ -n "$configured" ]; then
+    for index in $(cat $tmp_list | sed 's/\.gz$//g'); do
+      ghe_verbose "* Restoring $index"
+      echo "export PATH=\$PATH:/usr/local/share/enterprise && gzip -dc $GHE_REMOTE_DATA_USER_DIR/elasticsearch-restore/$index | ghe-es-load-json 'http://localhost:9201/$index'" |
+      ghe-ssh "$GHE_HOSTNAME" -- /bin/bash 1>&3
+    done
+  fi
+
+  ghe-ssh "$GHE_HOSTNAME" -- "sudo rm $GHE_REMOTE_DATA_USER_DIR/elasticsearch-restore/*.gz" 1>&3
```
Comment (snh) [inline share/github-backup-utils/ghe-restore-es-hookshot]: As before, doesn't this `rm` need to happen as part of the `if $CLUSTER || [ -n "$configured" ];` block?

File: share/github-backup-utils/ghe-restore-es-audit-log (hunk 1 above)
Comment (lildude) [reply share/github-backup-utils/ghe-restore-es-audit-log]: Yes and no. Yes cos that's about the only time it'll need to be run. No cos with a `-f` it'll work in all cases 😄

File: share/github-backup-utils/ghe-restore-es-hookshot (hunk 2 above)
Comment (lildude) [reply share/github-backup-utils/ghe-restore-es-hookshot]: As above.

File: share/github-backup-utils/ghe-restore-es-audit-log (hunk 1 above)
Comment (lildude) [reply share/github-backup-utils/ghe-restore-es-audit-log]: Actually, ignore me. You're right. Deleting it here would mean the logs are removed for unconfigured instances before they can be imported. Changing.

**Pull request merged**